### ISSUE TITLE
Proposal: Behavioral Configuration

### ISSUE THREAD

**Issue by SantoDE** [traefik/traefik#7550]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

This proposal is centered around a new configuration paradigm: a behavioral configuration
(in contrast to the current implementation, more declarative, yet powerful and flexible).

This behavioral configuration aims to simplify the setup of basic use cases.

With no further ado, this is how it may look for the file provider.

```yaml
# static configuration
entryPoints:
  web:
    address: :80

  websecure:
    address: :443
```

```yaml
# dynamic configuration
HTTPRoute: # maps to a combination of router/service
  rule: Host(`my.blog.com`) && (Path(`/...`) || ...)
  entrypoint: websecure
  redirectFrom: web
  tls:
    resolver: letsencrypt
    certificate: /path/to/cert # Path to Certificate
    key: /path/to/cert # Path to Certificate Key
  # Request modifier
  stripPrefix: 
    priority: 12 # all modifiers have a default priority that may be overriden
    prefixes:
      - foobar
      - foobar
    forceSlash: true
  ##
  loadBalancer:
    weighted:
      servers:
        - url: http://<private-ip-server-1>:<private-port-server-1>/
          weight: 1
        - url: http://<private-ip-server-2>:<private-port-server-2>/
```

Objects like HTTPRoute provide an opt-in to behavioral configuration patterns. For example, the declarative approach may look like:

```yaml
# static configuration
entryPoints:
  web:
    address: :80
      http:
        redirections:
          entryPoint:
            to: websecure
            scheme: https

  websecure:
    address: :443
```

```yaml
# dynamic configuration
http:
  # Add the routerHTTPouters:
  router0:
    rule: Host(`my.blog.com`) && (Path(`/...`) || ...) 
    entryPoints:
      - websecure
    middlewares:
      - my-stripPrefix
    service: service-foo
    tls: {}
  ##
  middlewares:
    test-stripprefix:
      stripPrefix:
        prefixes:
          - "/foobar"
          - "/foobar"
        forceSlash: true
  ##
  services:
    service-foo:
      loadBalancer:
        servers:
        - url: http://foo/
        - url: http://bar/
##
tls:
  certificates:
    - certFile: /path/to/cert # Path to Certificate
      keyFile: /path/to/cert # Path to Key for Certificate
```

In Docker Swarm, this structure may appear as:

```yaml
labels:
  - "traefik.httproute.rule=Host(`my.blog.com`) && (Path(`/...`) || ...)"
  - "traefik.httproute.entrypoint=websecure"
  - "traefik.httproute.redirectFrom=web"
  - "traefik.httproute.tls.resolver=letsencrypt" 
  - "traefik.httproute.cors.priority=42" # all middlewares have a default priority that could be overridden to keep a default behavior
  - "traefik.httproute.cors.accessControlAllowCredentials=true"
  - "traefik.httproute.cors.accessControlAllowMethods=foo,bar"
  - "traefik.httproute.cors.accessControlAllowOrigin=foobar"
  - "traefik.httproute.cors.accessControlMaxAge=42"
  - "traefik.httproute.stripPrefix.priority=42"
  - "traefik.httproute.stripPrefix.prefixes=foo,bar"
  - "traefik.httproute.stripPrefix.forceSlash=true"
```

We already see some limitations with this approach, e.g.:

* Not suited for complex scenarios (limited to 1 router per container, 1 middleware per type per router, ...)
* No middleware reference: you need to copy/paste the same configuration everywhere
* 1 file per router (TOML / YAML)
* load balancing limited to round-robin

To counterbalance the limitations, we're thinking about introducing this new system while keeping support for the existing one.
This would provide users with quick configuration options for common use cases (behavioral configuration)
while allowing for more complex scenarios (current configuration).

This is an open discussion with you, the community, and the Traefik maintainers, so please let us know what you think.
Together we can continue to further enhance Traefik and ensure it is always easy to use.

**Comment by ldez** [comment]
Another approach can be the approach that I suggested in https://github.com/traefik/traefik/issues/5630#issuecomment-593628496


Model/kind/default/template for routers:

```yml
# Dynamic configuration.
# We used File provider only the demo (it can be available on all providers)
http:
  kinds:
    kind0:
      router:
        entryPoints:
          mergePolicy: force
          value:
            - foobar
            - foobar
        middlewares:
          mergePolicy: force
          value:
            - foobar
            - foobar
        priority:
          mergePolicy: force
          value: 42
        tls:
          mergePolicy: force
          value: true
          options:
            mergePolicy: force
            value: foobar
          certResolver:
            mergePolicy: force
            value: foobar
          domains:
            mergePolicy: force
            values:
              - main: foo
                sans:
                  - foo
                  - foo
              - main: bar
                sans:
                  - bar
                  - bar
```

**Comment by AndrewSav** [comment]
I like @ldez option more. @SantoDE I'm afraid that supporting the OP is going to be a support nightmare. I understand that the goal is to make it simpler, but I can see potential for confusion by having two different but not exactly lining up ways of doing the same thing. Some people already struggle with static vs dynamic concept, etc... If it was me, I would not complicated this even further.

**Comment by emilevauge** [comment]
The template approach is certainly more generic/powerful, but it also implies more complexity.
The main idea behind this proposal is clearly to simplify the configuration for beginners. They should be able to set up common uses cases without learning 10 new Traefik concepts (static config, dynamic, entrypoints, routers, services, servers, middlewares, kind, etc). In one word, the idea is to ask beginners to learn as little as possible when it's for common use cases.
Templates could bring a default mechanism in traefik, but I'm not convinced it would simplify beginners' life that much.

**Comment by sqlsolver** [comment]
@emilevauge I'd like to offer a little of that feedback you're seeking. The reason it was a struggle for this beginner to differentiate between static and dynamic boils down to incomplete documentation. The only list of values I saw for reference is the [CLI static vars](https://doc.traefik.io/traefik/reference/static-configuration/cli/). If I saw a value and wondered if it would go in a dynamic configuration I referenced that list - if it wasn't there I guessed it could be used dynamically (the dynamic configuration sample was helpful for style, but not usage). A great cheat sheet might be a **two-column list** with static on one side, and dynamic on the other (shown without the CLI formatting as the new tabbed format examples are really good). 

Looking around the web you'll see quite a few articles simply trying to elucidate the documentation by explaining what Traefik is doing with a given value. I really do like the task-based approach of your docs but I think at least a **cross-referenced glossary** is needed, along with some more pedantic explanations. I may have read your entire library before I saw what PathPrefix is used. It would also clarify some of your product specific terminology. When I saw middlewares I didn't initially realize they were part of Traefik, I was thinking of specifying third-party middleware.

The other gotcha is that it looks like the repo has about 15% of its issues unresolved and some that are closed are aging out rather than being resolved. Things like #7414 which I commented on the other day mean that this otherwise very cool automation tool requires extra manual intervention for a typical process. That's fine for my POC, but I can't yet recommend it for an enterprise client. Meantime, **documenting known pitfalls** as the way it currently has to be done may reduce beginner frustration.

**Comment by pcfens** [comment]
One reason I initially adopted Traefik is because I was easily able to get started with it, and as my use cases got more complex I was able to add on what I needed (first stickiness, then middlewares, now plugins). I think that creating some threshold of complexity where users have to switch to a slightly different configuration will cause more problems than are solved (how is that threshold documented, and how do I decide which to use if I'm launching a new thing). The [Zen of Python](https://www.python.org/dev/peps/pep-0020/#id2) sums it up nicely: `There should be one-- and preferably only one --obvious way to do it.`

I think what made Traefik 1.x so easy wasn't as much the type of configuration, but rather the minimal and familiar nature of it. The 3 required components of port, frontend rule, and backend name are all concepts that sysadmins with pre-container experience will be familiar with, as well as some home users because of some parallels with NAT configuration.

There are a number of complex pieces of software that are relatively easy to get started with despite the need for extensive (and sometimes repetitive) configurations. [Istio](https://istio.io/latest/docs/setup/getting-started/#install), [Linkerd](https://linkerd.io/2/getting-started/#step-3-install-linkerd-onto-the-cluster), and the [Prometheus operator](https://github.com/prometheus-operator/prometheus-operator#quickstart) are all different but great approaches of hiding the complexity until it's needed by the user (something Traefik v1 did exceptionally well).

To summarize so far, a simple configuration is great, but don't lose sight of what actually makes for good software.

Finally, while the innovations created and enabled by Traefik are exciting, the day to day operations should remain boring. Traefik's role in a production environment is critical, and as such, users should expect some level of stability in the configuration, even if it's not perfect. That's not to say improvements shouldn't be made, but as Traefik finds itself gaining a more prominent place in the enterprise, it should be considered.

**Comment by soletan** [comment]
@pcfens I think Traefik isn't complying with this _one obvious way_ pattern already: there is toml and yaml and both formats come with different syntax rules to manage. In docker swarm I have to write rules in a totally different way. Unless I'm creating service descriptions each and every working day I'm far from getting sufficiently familiar with the existing syntax and thus rely on a self-made generator for creating those rule sets every time I'm deploying another micro service.

And due to its given limitations I don't expect this new syntax to replace the existing one but to be transformed into the latter internally just like there is syntactic sugar in many other languages nowadays. And IMHO reducing complexity is essential for making good software better. (e.g. that's why I'm having issues myself with getting deeply involved with k8s as it might serve any purpose and my imagination is the only (upper) limit there is, but on the other end it takes too much to achieve something as simple as getting a mostly stupid REST API up and running on bare metal. k8s still needs this optional reduction for people who don't need to see support for each and every opportunity there is.)

**Comment by gustavoninetyone** [comment]
Hello guys! Does anyone have news about this? I'm currently struggling to download a report in an external API. Every time I get the response a new cookie is created and my download and session is invalidated.

Example:
`domain.com/app/download?id...`

**Comment by nmengin** [comment]
Hello there,

This proposal has been open for a while with minimal traction. We won't be adding it to our roadmap at this time and prefer to close it for now.

If circumstances change or interest grows, we can always reopen this later.
Thank you, @SantoDE, for the suggestion!
